bin/propolis-server/src/lib/vm/request_queue.rs

```diff
@@ -164,248 +214,330 @@ enum RequestDisposition {
+                }
 
-            // Requests to stop always succeed. Note that a request to stop a VM
-            // that hasn't started should still be queued to the state worker so
-            // that the worker can exit and drop its references to the instance.
-            ExternalRequest::Stop => self.allowed.stop,
+                // Always enqueue component change requests, even if the VM has
+                // a pending request to stop or migrate out. This allows the
+                // state driver to process these requests during a state change,
+                // which may be necessary to complete that state change. If the
+                // change request is canceled by a later state transition, the
+                // queue can use the request data to notify the requestor.
+                ExternalRequest::Component(
+                    ComponentChangeRequest::ReconfigureCrucibleVolume {
+                        ..
+                    },
+                ) => Disposition::Enqueue,
+            }
         };
 
         info!(&self.log, "Queuing external request";
               "request" => ?request,
               "disposition" => ?disposition);
 
         match disposition {
-            RequestDisposition::Enqueue => {}
-            RequestDisposition::Ignore => return Ok(()),
-            RequestDisposition::Deny(reason) => return Err(reason),
-        };
-
-        self.allowed = self.get_new_dispositions(
-            DispositionChangeReason::ApiRequest(&request),
-        );
-        self.queue.push_back(request);
-        Ok(())
-    }
-
-    /// Notifies the queue that the instance's state has changed and that its
-    /// disposition should be updated accordingly.
-    pub fn notify_instance_state_change(&mut self, state: InstanceStateChange) {
-        self.allowed = self
-            .get_new_dispositions(DispositionChangeReason::StateChange(state));
-    }
-
-    /// Computes a new set of queue dispositions given the current state of the
-    /// queue and the event that is changing those dispositions.
-    fn get_new_dispositions(
-        &self,
-        reason: DispositionChangeReason,
-    ) -> AllowedRequests {
-        debug!(self.log, "Computing new queue dispositions";
-               "reason" => ?reason);
-
-        use DispositionChangeReason as ChangeReason;
-        use RequestDeniedReason as DenyReason;
-        use RequestDisposition as Disposition;
-        match reason {
-            ChangeReason::ApiRequest(ExternalRequest::Start) => {
-                let reason = DenyReason::StartInProgress;
-                AllowedRequests {
-                    start: Disposition::Ignore,
-                    migrate_as_source: Disposition::Deny(reason),
-                    reboot: Disposition::Deny(reason),
-                    mutate: Disposition::Deny(reason),
-                    stop: self.allowed.stop,
+            Disposition::Ignore => Ok(()),
+            Disposition::Deny(reason) => Err(reason),
```

Comment:
soooo...i'm not convinced that the disposition enum is really pulling its weight anymore. Why not just have the arms of the above match that presently return `Disposition::Ignore` and `Disposition::Deny` just `return Ok(())`/`return Err(reason)`? i guess the `Disposition`s do get logged in one place, which is maybe nicer than having all the places where we would return log separately...

Reply:
Done via 29ea5fc and 0178609.